File: backend/services/docxGenerator.js
import {
  Document,
  Packer,
  Paragraph,
  TextRun,
  HeadingLevel,
  AlignmentType,
  BorderStyle
} from "docx";

export async function generateResumeDocx(resume) {
  const pInfo = resume.personalInfo || {};
  const workExp = resume.workExperience || [];
  const edu = resume.education || [];
  const skills = resume.skills || {};
  const projects = resume.projects || [];
  const certifications = resume.certifications || [];

  const children = [];

  // 1. Header — Name & Contact
  children.push(
    new Paragraph({
      alignment: AlignmentType.CENTER,
      children: [
        new TextRun({
          text: pInfo.fullName || "Candidate Name",
          bold: true,
          size: 32, // 16pt
          font: "Arial"
        })
      ],
      spacing: { after: 120 }
    })
  );

  const contactPieces = [];
  if (pInfo.email) contactPieces.push(pInfo.email);
  if (pInfo.phone) contactPieces.push(pInfo.phone);
  if (pInfo.location) contactPieces.push(pInfo.location);
  if (pInfo.linkedin) contactPieces.push(pInfo.linkedin);
  if (pInfo.github) contactPieces.push(pInfo.github);
  if (pInfo.portfolio) contactPieces.push(pInfo.portfolio);

  if (contactPieces.length > 0) {
    children.push(
      new Paragraph({
        alignment: AlignmentType.CENTER,
        children: [
          new TextRun({
            text: contactPieces.join(" | "),
            size: 20, // 10pt
            font: "Arial"
          })
        ],
        spacing: { after: 240 }
      })
    );
  }

  const createSectionHeader = (title) => {
    return new Paragraph({
      heading: HeadingLevel.HEADING_2,
      border: {
        bottom: {
          color: "1A91F0",
          space: 2,
          style: BorderStyle.SINGLE,
          size: 12
        }
      },
      children: [
        new TextRun({
          text: title.toUpperCase(),
          bold: true,
          size: 24, // 12pt
          color: "1A91F0",
          font: "Arial"
        })
      ],
      spacing: { before: 200, after: 120 }
    });
  };

  // 2. Summary
  if (resume.summary) {
    children.push(createSectionHeader("Professional Summary"));
    children.push(
      new Paragraph({
        children: [
          new TextRun({
            text: resume.summary,
            size: 21,
            font: "Arial"
          })
        ],
        spacing: { after: 160 }
      })
    );
  }

  // 3. Technical Skills
  const techList = skills.technical || [];
  const toolsList = skills.tools || [];
  const softList = skills.soft || [];

  if (techList.length > 0 || toolsList.length > 0 || softList.length > 0) {
    children.push(createSectionHeader("Skills & Competencies"));

    if (techList.length > 0) {
      children.push(
        new Paragraph({
          children: [
            new TextRun({ text: "Technical Skills: ", bold: true, size: 21, font: "Arial" }),
            new TextRun({ text: techList.join(", "), size: 21, font: "Arial" })
          ],
          spacing: { after: 60 }
        })
      );
    }

    if (toolsList.length > 0) {
      children.push(
        new Paragraph({
          children: [
            new TextRun({ text: "Tools & Frameworks: ", bold: true, size: 21, font: "Arial" }),
            new TextRun({ text: toolsList.join(", "), size: 21, font: "Arial" })
          ],
          spacing: { after: 60 }
        })
      );
    }

    if (softList.length > 0) {
      children.push(
        new Paragraph({
          children: [
            new TextRun({ text: "Core Competencies: ", bold: true, size: 21, font: "Arial" }),
            new TextRun({ text: softList.join(", "), size: 21, font: "Arial" })
          ],
          spacing: { after: 160 }
        })
      );
    }
  }

  // 4. Experience
  if (workExp.length > 0) {
    children.push(createSectionHeader("Work Experience"));

    for (const exp of workExp) {
      children.push(
        new Paragraph({
          children: [
            new TextRun({ text: exp.role || "Role", bold: true, size: 22, font: "Arial" }),
            new TextRun({ text: ` — ${exp.company || "Company"}`, bold: true, size: 22, font: "Arial", color: "444444" }),
            new TextRun({
              text: `\t${exp.startDate || ""} - ${exp.current ? "Present" : exp.endDate || ""}`,
              size: 20,
              font: "Arial",
              color: "666666"
            })
          ],
          spacing: { before: 100, after: 60 }
        })
      );

      for (const bullet of exp.bullets || []) {
        if (bullet && bullet.trim()) {
          children.push(
            new Paragraph({
              bullet: { level: 0 },
              children: [
                new TextRun({
                  text: bullet.trim(),
                  size: 21,
                  font: "Arial"
                })
              ],
              spacing: { after: 40 }
            })
          );
        }
      }
    }
  }

  // 5. Projects
  if (projects.length > 0) {
    children.push(createSectionHeader("Projects"));

    for (const proj of projects) {
      const techStr = proj.techStack && proj.techStack.length > 0 ? ` | Tech: ${proj.techStack.join(", ")}` : "";
      children.push(
        new Paragraph({
          children: [
            new TextRun({ text: proj.title || "Project", bold: true, size: 22, font: "Arial" }),
            new TextRun({ text: techStr, size: 20, font: "Arial", color: "666666" })
          ],
          spacing: { before: 80, after: 40 }
        })
      );

      if (proj.description) {
        children.push(
          new Paragraph({
            bullet: { level: 0 },
            children: [
              new TextRun({
                text: proj.description,
                size: 21,
                font: "Arial"
              })
            ],
            spacing: { after: 40 }
          })
        );
      }
    }
  }

  // 6. Education
  if (edu.length > 0) {
    children.push(createSectionHeader("Education"));

    for (const item of edu) {
      children.push(
        new Paragraph({
          children: [
            new TextRun({
              text: `${item.degree || "Degree"} ${item.field ? `in ${item.field}` : ""}`,
              bold: true,
              size: 22,
              font: "Arial"
            }),
            new TextRun({ text: ` — ${item.institution || "Institution"}`, size: 22, font: "Arial" }),
            new TextRun({
              text: `\t${item.year || ""}${item.grade ? ` | ${item.grade}` : ""}`,
              size: 20,
              font: "Arial",
              color: "666666"
            })
          ],
          spacing: { before: 60, after: 60 }
        })
      );
    }
  }

  // 7. Certifications
  if (certifications.length > 0) {
    children.push(createSectionHeader("Certifications"));
    for (const cert of certifications) {
      children.push(
        new Paragraph({
          bullet: { level: 0 },
          children: [
            new TextRun({ text: cert, size: 21, font: "Arial" })
          ],
          spacing: { after: 40 }
        })
      );
    }
  }

  const doc = new Document({
    sections: [
      {
        properties: {
          page: {
            margin: {
              top: 720, // 0.5 inch
              right: 720,
              bottom: 720,
              left: 720
            }
          }
        },
        children
      }
    ]
  });

  return await Packer.toBuffer(doc);
}

File: backend/app/services/docx_generator.py
"""DOCX export — structured resume → ATS-friendly Word document.

Layer C of the source repository's rule system forbids tables, columns,
images, and special formatting in the final ATS-safe resume. We obey
that here.

Layer A forbids emojis and special characters in user-visible text.
"""

from __future__ import annotations

import io
from typing import Any

from docx import Document
from docx.enum.text import WD_ALIGN_PARAGRAPH
from docx.shared import Pt, RGBColor

from app.engine.schemas import ResumeStructured


def render_resume_to_docx(resume: ResumeStructured) -> bytes:
    """Build a .docx file from a structured resume and return it as bytes."""
    doc = Document()

    # Tight margins for ATS readability (Layer C reference).
    for section in doc.sections:
        section.top_margin = Pt(36)
        section.bottom_margin = Pt(36)
        section.left_margin = Pt(54)
        section.right_margin = Pt(54)

    _set_base_style(doc)

    _render_header(doc, resume)
    _render_summary(doc, resume)
    _render_skills(doc, resume)
    _render_experience(doc, resume)
    _render_education(doc, resume)
    if resume.languages:
        _render_languages(doc, resume)
    if resume.optional_sections:
        for sec in resume.optional_sections:
            _render_optional(doc, sec.title, sec.items)

    buf = io.BytesIO()
    doc.save(buf)
    return buf.getvalue()


# ---------- Section renderers ----------


def _set_base_style(doc: Document) -> None:
    style = doc.styles["Normal"]
    style.font.name = "Calibri"
    style.font.size = Pt(11)


def _render_header(doc: Document, resume: ResumeStructured) -> None:
    h = resume.header

    name_para = doc.add_paragraph()
    name_para.alignment = WD_ALIGN_PARAGRAPH.CENTER
    name_run = name_para.add_run(h.full_name or "")
    name_run.bold = True
    name_run.font.size = Pt(18)

    title_para = doc.add_paragraph()
    title_para.alignment = WD_ALIGN_PARAGRAPH.CENTER
    title_run = title_para.add_run(h.title or "")
    title_run.font.size = Pt(12)
    title_run.italic = True

    if h.contacts:
        contact_para = doc.add_paragraph()
        contact_para.alignment = WD_ALIGN_PARAGRAPH.CENTER
        contact_run = contact_para.add_run(" | ".join(h.contacts))
        contact_run.font.size = Pt(10)
        contact_run.font.color.rgb = RGBColor(0x55, 0x55, 0x55)


def _section_heading(doc: Document, text: str) -> None:
    para = doc.add_paragraph()
    run = para.add_run(text.upper())
    run.bold = True
    run.font.size = Pt(12)
    run.font.color.rgb = RGBColor(0x1A, 0x1C, 0x6A)


def _render_summary(doc: Document, resume: ResumeStructured) -> None:
    if not (resume.summary and resume.summary.summary_text):
        return
    _section_heading(doc, "Summary")
    doc.add_paragraph(resume.summary.summary_text)


def _render_skills(doc: Document, resume: ResumeStructured) -> None:
    if not resume.skills.groups:
        return
    _section_heading(doc, "Skills")
    for grp in resume.skills.groups:
        if not grp.items:
            continue
        line = grp.group_name + ": " + ", ".join(grp.items)
        doc.add_paragraph(line)


def _render_experience(doc: Document, resume: ResumeStructured) -> None:
    if not resume.experience:
        return
    _section_heading(doc, "Experience")
    for exp in resume.experience:
        head = doc.add_paragraph()
        role_run = head.add_run(f"{exp.role} — {exp.company}")
        role_run.bold = True
        head.add_run(f"   {exp.start_date} – {exp.end_date}").italic = True
        for h in exp.highlights:
            bullet = doc.add_paragraph(h, style="List Bullet")
            bullet.paragraph_format.left_indent = Pt(18)


def _render_education(doc: Document, resume: ResumeStructured) -> None:
    if not resume.education:
        return
    _section_heading(doc, "Education")
    for ed in resume.education:
        line = f"{ed.institution} — {ed.degree}"
        if ed.field:
            line += f", {ed.field}"
        para = doc.add_paragraph()
        run = para.add_run(line)
        run.bold = True
        if ed.start_date or ed.end_date:
            para.add_run(f"   {ed.start_date or ''} – {ed.end_date or ''}").italic = True
        if ed.grade:
            doc.add_paragraph(f"Grade: {ed.grade}")


def _render_languages(doc: Document, resume: ResumeStructured) -> None:
    if not resume.languages:
        return
    _section_heading(doc, "Languages")
    doc.add_paragraph(
        ", ".join(f"{l.name} ({l.level})" for l in resume.languages)
    )


def _render_optional(doc: Document, title: str, items: list[str]) -> None:
    if not items:
        return
    _section_heading(doc, title)
    for item in items:
        doc.add_paragraph(item, style="List Bullet")
